### ISSUE TITLE
Interaction with features

### ISSUE THREAD

**Issue by hactar** [stadiamaps/maplibre-swiftui-dsl-playground#18]
Hi there, starting a further discussion here. 

I'd need to implement basic interaction with features. So in the non-dsl world I'd add a gesture recognizer to the mapview and then I'd usually run something like

```Swift
// calculate the rect based on tap location
let features = mapView.visibleFeatures(in: rect, styleLayerIdentifiers: tappableLayers)
guard let feature = features.first else {
   selected(nil)
   return
}

// use the feature.attributes to match what I'm looking for
if let placeID = feature.attribute(forKey: "mapItemId") as? String {
  selected(placeID)   
}
```

How are we planning to support interacting with features of the map?

SwiftUI has things like `.onTapGesture` for views. Is this the path for this dsl too? My first thought would be:

```Swift
MapView()
  .onTapGesture(layers: ["layerID1", "layerID2"] { features in
    // visibleFeatures runs for [layerID1, layerID2], returning an optional array of visibleFeatures that were tapped
}
```

Interested in hearing your thoughts - I might be able to work on a PR for this.

**Comment by ianthetechie** [comment]
Short answer, yes, that’s the general path that I was imagining. I believe it *might* be possible to improve the safety of the API though. 

As you know, everything in MapLibre is “stringly” typed with respect to these sorts of layer references. What I’d like to explore is if we can move this to being a modifier on the layer or something that involves an actual layer reference so we don’t have to tie things together with strings.

I haven’t given this a ton of thought, so counter proposals are definitely welcome.

We should also have a look at MapKit’s latest API to see if we can borrow any ideas. (MapKit is quite a bit different of course so it might not be possible but worth a look. I seem to remember them doing a lot with adding individual features and making them interactive, which is honestly an easier model for most developers to reason about, so it would be great if we could also enable this eventually).

**Comment by hactar** [comment]
Yes, the non type-safe approach isn't exactly great. I usually at least create myself an emum of strings for layer names to mitigate the issue, but its not a perfect solution.

I hadn't considered attaching the tap gesture to the layer instead of the view. I do like the idea as it would take a lot of the filtering responsibility away from users and removes a number of mistakes developers could make. Two questions/difficulties arise though:

- Under the hood i picture this code to be quite complex. For performance reasons we probably still just want to run  `visibleFeatures` once per tap, and not once per tap per layer. So we would have to maintain our own mapping on which closure is listening for which layer, and return the visible features to the correct closure - sometimes with multiple closures being triggered by one tap.
- Some developers may want to make a layer tappable that they did not create themselves - one that is baked into the style.json file. In this case we would still need to provide a string based approach on the MapView level?

MapKit thinks less of gestures, and more in terms of selection. You get a binding of what the user selected. I like that too - it is obviously way more limited though in what you can do with it - at the same time it probably would cover what most devs want to do in a way simpler form. Apple also has a MapReader with a proxy for the more advanced tapping.

```Swift
struct MarkerSelectionView: View {
    @State private var selectedTag: Int?
    
    var body: some View {
        Map(selection: $selectedTag) {
            Marker("Empire state building", coordinate: .empireStateBuilding)
                .tint(.orange)
                .tag(1)
            
            Annotation("Columbia University", coordinate: .columbiaUniversity) {
                ZStack {
                    RoundedRectangle(cornerRadius: 5)
                        .fill(selectedTag == 2 ? .red : .teal)
                    Text("Something")
                        .padding(5)
                }
            }
            .tag(2)
        }
    }
}
```

**Comment by hactar** [comment]
Closing this as implemented via #30
